File: backend-nestjs/src/modules/damage-reports/application/commands/create-damage-report/create-damage-report.handler.ts
import { CommandHandler, ICommandHandler, EventBus } from '@nestjs/cqrs';
import { CreateDamageReportCommand } from './create-damage-report.command';
import { DamageReportRepository } from '../../../domain/damage-report.repository';
import { DamageReport } from '../../../domain/damage-report';
import { DamageReportCreatedEvent } from '../../events/damage-report-created.event';

@CommandHandler(CreateDamageReportCommand)
export class CreateDamageReportHandler implements ICommandHandler<
  CreateDamageReportCommand,
  DamageReport
> {
  constructor(
    private readonly repository: DamageReportRepository,
    private readonly eventBus: EventBus,
  ) {}

  async execute(command: CreateDamageReportCommand): Promise<DamageReport> {
    const report = DamageReport.create({
      zoneId: command.zoneId,
      type: command.type,
      severity: command.severity,
      lat: command.lat,
      lng: command.lng,
      description: command.description,
      reportedBy: command.reportedBy,
    });

    const saved = await this.repository.save(report);

    this.eventBus.publish(
      new DamageReportCreatedEvent(
        saved.id,
        saved.zoneId,
        saved.type,
        saved.severity,
        saved.description,
      ),
    );

    return saved;
  }
}

File: backend-springboot/src/main/java/com/greenalgeria/damagereport/application/command/CreateDamageReportHandler.java
package com.greenalgeria.damagereport.application.command;

import com.greenalgeria.damagereport.application.*;
import com.greenalgeria.damagereport.domain.DamageReport;
import com.greenalgeria.damagereport.domain.DamageReportRepository;
import com.greenalgeria.shared.cqrs.CommandHandler;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

@Component
@Transactional
public class CreateDamageReportHandler implements CommandHandler<CreateDamageReportCommand, DamageReportResponse> {

    private final DamageReportRepository damageReportRepository;

    public CreateDamageReportHandler(DamageReportRepository damageReportRepository) {
        this.damageReportRepository = damageReportRepository;
    }

    public DamageReportResponse handle(CreateDamageReportCommand command) {
        var request = command.request();
        var report = new DamageReport(
                request.zoneId(),
                request.type(),
                request.severity(),
                request.lat(),
                request.lng(),
                request.description(),
                request.reportedBy());
        var saved = damageReportRepository.save(report);
        return DamageReportResponse.from(saved);
    }

    @Override
    public Class<CreateDamageReportCommand> supportedCommand() {
        return CreateDamageReportCommand.class;
    }
}
